File: SearchInRotatedSortedArray.java
public class SearchInRotatedSortedArray {
    public int search(int[] arr, int target) {
    
		int pivot = getPeak(arr);
        if(pivot==-1)
            return binarySearch(arr,target,0,arr.length-1);
		int index = binarySearch(arr, target, 0, pivot);
		if(index==-1)
			index=binarySearch(arr, target, pivot+1, arr.length-1);
		return index;
    }
    public int binarySearch(int arr[], int target, int start, int end) {
		while(start<=end) {
			int mid=start + (end-start)/2;
			if(arr[mid]<target) {
				start=mid+1;
			}
			else if(arr[mid]>target) {
				end=mid-1;
			}
			else {
				return mid;
			}
		}
		return -1;
	}
	public int getPeak(int []arr){
		int start=0;
		int end = arr.length-1;
		while(start<end) {
			int mid = start + (end-start)/2;
			if(mid<end && arr[mid]>arr[mid+1])
				return mid;
			if(mid>start && arr[mid]<arr[mid-1])
				return mid-1;
			if(arr[mid]<= arr[start]) {
				end=mid-1;
			}
			else if(arr[mid]>= arr[start]) {
				start=mid+1;
			}
		}
		return -1;
	}

    public static void main(String[] args) {
        int []array = {3,1};
        SearchInRotatedSortedArray searchInRotatedSortedArray = new SearchInRotatedSortedArray();
        int sol = searchInRotatedSortedArray.search(array, 1);
        System.out.println(sol);
    }
}

File: searchInRotatedSortedArray.go
package main

import "fmt"

func searchInRotatedSortedArray(nums []int, target int) int { //func name changed
	pivotIndex := findPivot(nums)

	first := binarySearch(nums, 0, pivotIndex-1, target)
	if first != -1 {
		return first
	}
	second := binarySearch(nums, pivotIndex, len(nums)-1, target)

	return max(first, second)

}

// to find pivot always use a variable
func findPivot(nums []int) int {
	start := 0
	end := len(nums) - 1
	minimum := 0
	for start <= end {
		mid := start + (end-start)/2
		if nums[minimum] > nums[mid] {
			minimum = mid
		}
		if nums[mid] > nums[end] {
			start = mid
		} else {
			end = mid - 1
		}
	}
	return minimum

}

func binarySearch(nums []int, start int, end int, target int) int {
	if start > end {
		return -1
	}
	mid := start + (end-start)/2

	if nums[mid] == target {
		return mid
	}

	if nums[mid] > target {
		return binarySearch(nums, start, mid-1, target)
	} else {
		return binarySearch(nums, mid+1, end, target)
	}
}

func main() {
	fmt.Println(searchInRotatedSortedArray([]int{5, 1, 3}, 1))
}
